### PR TITLE
[COMPASS-1680] Feature: Webpack Build

### PR DESCRIPTION
**Pull request by matt-d-rat**
# Summary

This pull request updates the compass-plugin template to use [Webpack](https://webpack.js.org/) to build compass plugins.

Tests are also run with Webpack (via by [mocha-webpack](http://zinserjan.github.io/mocha-webpack/) so that we can be certain that the code under test, is the same as the output as is for builds in production or development. Code coverage is also provided by Istanbul via the [NYC CLI](https://github.com/istanbuljs/nyc).

[CSS modules](https://github.com/css-modules/css-modules) are enabled to provide scoped CSS per component. The global CSS file, located in `src/assets/less/index.less` is/should ideally be the only file for declaring global CSS - this is what imports the global styles used by compass currently.

## Commands

- `npm start` - Start a standalone electron app with a [webpack-dev-server](https://webpack.js.org/configuration/dev-server/) and [hot module reloading (HMR)](https://webpack.js.org/concepts/hot-module-replacement/) (with [react-hot-loader](https://github.com/gaearon/react-hot-loader)) support!

- `npm run start:prod` - Start a standalone electron app and run the compiled output in the `lib` folder. Useful for sanity testing the prod build external to compass.

- `npm run start:watch` - Watch the file system for changes and re-compile the output in the `lib` folder (for dev). Used as a stop gap when developing plugins inside Compass until such time as Compass uses Webpack itself. _Note: this is slower than running standalone dev (`npm start`) and does not support HMR._

- `npm run compile` - Build the plugin for production use.

- `npm test` - Run the tests.

- `npm run test:watch` - Run the tests in watch mode.

- `npm run cover` - Run the tests and generate a code coverage report.

## Screenshots

plugin-test-webpack-dev

plugin-test-webpack-prod

plugin-test-webpack-compass

plugin-test-runtest

# Running this branch yourself

If you want to try out these changes for yourself on your own machine follow the instructions below to generate a plugin:

```
git clone git@github.com:mongodb-js/compass-plugin.git
cd compass-plugin
git pull
git checkout feature/webpack-build
cd ..
khaos create -d ./ compass-plugin ./plugin-test
```
Then follow the on screen instructions from the Khaos template.

# Notes

## Structure Changes

- `src/index.js` - The entry point to the plugin, exposes the Connected Plugin, App Registry activate and deactivate functions (Formerly located in the root directory).
- `src/Plugin.js` - The Connected Plugin, wraps the root React Component with the `StoreConnector
- `src/components/{{name}}` - The root React Component for the compass plugin

- `src/components` - All components now live under there own folder (camelcase) which contains the following files:
    - `index.js` - Entry for the component, exposing ES6 default and named exports.
    - `{{name}}.jsx` - The React Component
    - `{{name}}.less` - The scoped styles for this component (see [CSS Modules](https://github.com/css-modules/css-modules))
    - `{{name.spec.js}}` -  The test file for this component

- `config` - Contains the configuration files for the project and webpack.

- `test` - Now only contains the `setup.js` which configures a number of globals for the mocha test runner and sets up a jsdom environment. Unit and Enzyme tests now live next to the code under test instead of a seperate folder.

## Webpack Aliases

You may notice through out the code import/require statements written in this form `import foo from 'components/foo'` or `@import '~less/compass/index.less'`. These imports are not relative paths, rather they use [Webpack Resolve Aliases](https://webpack.js.org/configuration/resolve/) to resolve the paths to the files. See Webpack config for details.

## CSS Modules + Compass

The generated index.css file in `lib/assets/css/index.css` is only used by the standalone prod build (run with `npm run start:prod`) to provide the same set of global styles that would be provided by compass in this context. This index.css is __NOT__ used/imported by compass and should therefore not be relied upon by plugins for there own styles. Instead plugins should declare their own styles scoped per component (see `src/components/ToggleButton` for an example) for CSS modules, or alternatively add the global styles directly to compass.

# TODO

- [x] Add an npm command (`npm run start:watch`) to run webpack with just file system watcher so that working on internal compass plugins write out to the lib folder in dev mode on change so changes can be verified in compass (not as good as running as dev server as this is slower) - A stop gap until we have compass itself running with webpack
- [x] Storybook needs updating to use our custom webpack config so that components can still be developed in isolation from one another
- [ ] Need to verify that the link scripts still work, or alternatively explore an alternative approach (npm peerDependencies may be enough to satisfy this requirement)
- [ ] Update the compass-internal-docs to reflect these changes to the plugin template.
- [ ] Document and/or write a bash script upgrade path for existing plugins to use this new structure and tooling

### CODE REVIEW

File: template/electron/index.js [hunk 1]
```diff
@@ -1 +1,78 @@
...
+const path = require('path')
+const url = require('url')
+
+// Keep a global reference of the window object, if you don't, the window will
+// be closed automatically when the JavaScript object is garbage collected.
+let mainWindow;
+
+// Keep a reference for dev mode
+let dev = false;
+if ( process.defaultApp || /[\\/]electron-prebuilt[\\/]/.test(process.execPath) || /[\\/]electron[\\/]/.test(process.execPath) ) {
```
Comment (pzrq) [inline template/electron/index.js]: It is my understanding that [`electron-prebuilt` was renamed to just `electron` in npm](https://www.npmjs.com/package/electron#installation) quite a while ago (i.e. Compass v1.7 is the last artifact we created that included it, and 1.8+ which was used for the plugins workshop was Electron 1.6.1).

Is there a reason to support the older process path?

File: template/electron/renderer/index.js [hunk 2]
```diff
@@ -1,38 +1,55 @@
...
-  // global.hadronApp.dataService = ds;
-  // global.hadronApp.appRegistry.onConnected(error, ds);
-// });
-
-ReactDOM.render(
-  React.createElement({{pascalcase name}}Component),
-  document.getElementById('container')
-);
+
+// Activste our plugin with the Hadron App Regustry
```
Comment (pzrq) [inline template/electron/renderer/index.js]: typo :)

**Comment by rueckstiess** [comment]
This is great @matt-d-rat. I think overall a massive win! 

For the review, let's collect issues we notice in our code reviews here and address over the next day or two. Then I would say we merge to unblock you from converting an existing plugin into the same structure, and if other smaller issues are found we can create separate PRs.

First up: the linter is complaining. `npm run check` runs the eslint tests. There are 33 problems. I won't list them all here, but they should be addressed. mostly style guide violations that can be fixed quickly, like indentations, missing semicolons, etc.

**Comment by rueckstiess** [comment]
Another thing that is introduced here that we haven't been using before is the `classnames` module. It looks complicated at first, but is actually quite simple, and as Matt told me, just read the manual: https://www.npmjs.com/package/classnames

Especially when using with React, this makes code a lot cleaner, see this example: https://www.npmjs.com/package/classnames#usage-with-reactjs

File: template/electron/index.js (hunk 1 above)
Comment (matt-d-rat) [inline template/electron/index.js]: I don't think it hurts, was more of a safety check given I didn't have that information available to me at the time.

File: template/electron/renderer/index.js (hunk 2 above)
Comment (matt-d-rat) [inline template/electron/renderer/index.js]: [ ] Fix typos

**Comment by satyasinha** [comment]
Not related to this PR but khaos doesn't work well in Windows. Running the create command gives an ENOENT error. There is an open PR to fix that [here](https://github.com/segmentio/khaos/pull/90) but it has been up for more than a year now.

Applying the above fix doesn't quite work because filenames aren't being changed from `{{name}}` version to the name provided during create.

This is a khaos issue not a webpack issue, which would work fine in Windows.

**Comment by matt-d-rat** [comment]
@satyasinha yeah not an issue for this PR or our project but worth noting. A manual solution would be to simply clone the project and copy the files directly out of template and do a find replace for the various Khaos placeholders. Failing that we could arguably dockerise the build, might be overkill though.

**Pull request merged**